tutorial.md

```diff
@@ -0,0 +1,225 @@
+SymPy is a pure Python library for symbolic mathematics. Code generation refers
+to the act of converting a SymPy symbolic expression into equivalent code in
+some language. This allows one to use SymPy to symbolically model a problem,
+and generate fast numerical code for specific platforms that executes that
+model. This is a powerful tool that is useful to scientists in many domains.
+Code generation allows users to speed up existing code, to deal only with the
+high level mathematics of a problem, avoids mathematical errors and typos,
+makes it possible to deal with expressions that would otherwise be too large to
+write by hand, and opens possibilities to perform automatic mathematical
+optimizations of expressions.
+
+SymPy supports generating code for C, C++, Fortran, Matlab/Octave, Python,
+Cython, Julia, Javascript, LLVM, Rust, Haskell, Mathematica, Tensorflow, and
+Theano, and can easily be extended to other languages. SymPy’s code generation
+is used by libraries such as PyDy, pyodesys, sympybotics, pycalphad, and many
+other programs.
+
+## Learning objectives
+
+Attendees will be able to:
+- write SymPy expressions describing mathematical functions and identify the
+  function arguments and outputs.
+- use the SymPy code printers to transform SymPy expressions representing
+  common domain specific functions into multiple output languages.
+- use the SymPy code generation routines to output compilable C code and use
+  Cython to access these functions in Python.
+- generate custom vectorized functions with the three SymPy functions:
+  lambdify, ufuncify, and autowrap.
+- create both custom code printers that make use of specialized C libraries and
+  common subexpression elimination (CSE).
+- subclass the core SymPy printers and create a printer for a custom language.
+
+## Outline
+
+Intro to SymPy Expressions [30 minutes]
+
+- Description: Writing common domain specific mathematical expressions with
+  SymPy.
+- Motivating Examples: Long expressions, Matrix operations, and Loop Fusion
+  from classical mechanics, chemical kinetics, nuclear dynamics, and materials
+  science.
+
+Code Printers [30 minutes]
+
+- Description: Printing expressions in multiple languages (C, Fortran, Rust,
+  Julia, Octave, Javascript, etc)
+- Motivating Example: 2D interactive plot in a Jupyter notebook by javascript
+  injection
+
+The Easy Way: High Level Generation (lambdify, ufuncify) [30 minutes]
+
+- Description: Generate loop fused NumPy ufuncs and compare to automatically
+  generated NumPy code. Show how you can extend lambdify with custom Python
+  functions.
+- Motivating Example: Generate a Jacobian function for a chemical kinetic
+  problem.
+
+The Harder Way: Code generation and compilation [1 hour]
+
+- Description: Write a tight low level loop with an indexed type and a long
+  expressions with knowing C.
+- Motivating Example: Evaluate the chemical kinetic Jacobian in a loop.
```

Comment:
The jacobian evaluation function is called intermittently by the solver. If we want a tight loop we could look at how one can use the method of lines to solve a PDE (e.g. diffusion - a.k.a. the heat equation). Then the jacobian would be evaluated at each space coordinate. (i.e. in a loop).